### ISSUE TITLE
[BUG] Noticed the following two error msgs occasionally in my server log

### ISSUE THREAD

**Issue by mikeysas** [AF-1/lms-customskip#1]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Check common causes of LMS (plugin) problems

- [X] The list has no fix for my issue

### What happened?

I have noticed any specific impact in usage other than seeing the following error messages in my server.log starting on 12/18.

`[24-12-18 18:16:16.8752] Slim::Control::Request::notify (2056) Error: Failed notify: Can't call method "id" on an undefined value at /var/lib/squeezeboxserver/cache/InstalledPlugins/Plugins/CustomSkip3/Plugin.pm line 1143.
[24-12-18 22:23:28.1771] Slim::Control::Request::notify (2056) Error: Failed notify: Can't call method "id" on an undefined value at /var/lib/squeezeboxserver/cache/InstalledPlugins/Plugins/CustomSkip3/Plugin.pm line 1130.`

At that time I was running:
`v9.0.1, 1734365592, Tue Dec 17 03:49:16 UTC 2024) perl 5.030000 - x86_64-linux-gnu-thread-multi`

And I have seen the first error again after upgrading to:
`v9.0.1, 1734552353, Thu Dec 19 03:00:58 UTC 2024) perl 5.030000 - x86_64-linux-gnu-thread-multi`

### Expected Behavior

No errors?

### Steps To Reproduce

Not sure?  It is random which I understand might be a challenge.

### LMS version

v9.0.1, 1734552353

### Plugin version

Custom Skip 3.6.7

### Your OS (incl. version)

Ubuntu 20.04

### What web skins / controllers are you seeing the problem with?

Material

### What browsers are you seeing the problem on?

Chrome

### Environment

```markdown
Version: 9.0.1 - 1734552353 @ Thu Dec 19 03:00:58 UTC 2024
Operating system: Debian - EN - utf8
Platform Architecture: x86_64-linux
Perl Version: 5.30.0 - x86_64-linux-gnu-thread-multi
Audio::Scan: 1.02
IO::Socket::SSL: 2.067
Database Version: DBD::SQLite 1.58 (sqlite 3.22.0)

Plugins

1001 Albums You Must Hear Before You Die 1.4.2
Additional Browse Modes 1.0
Advanced Tag View 1.0
Alternative Play Count 1.6.0
Bliss Mixer 0.4.0
Client Cleanup 1.2.0
Community Firmware for Touch/Radio/Controller 0.0.9
Context Stats 1.0.1
Custom Skip 3.6.7
Custom Tag Importer 1.2.4
Date and Time 1.0
Denon/Marantz AVR Control 5.3.1
Don't Stop The Music 1.0.0
Drag & drop music files to play 0.1
Dynamic Playlist Creator 1.6.7
Dynamic Playlists 4.5.0
Full text search 1.0.0
Group Players 0.17.2
Info Browser 1.0
Last.fm Audioscrobbler 1.0
LastMix 2.3.0
Line In 1.0
Line Out 1.0
Material Skin 5.7.0
Music and Artist Information 1.12.2
Music Information Screen 4.5.1
Music, Walk With Me 1.1.3
Network Test 1.0
Online Music Library Integration 1.0
piCorePlayer 0.1.2
PlayHLS 2.12
Podcasts 2.0
PotPourri 1.0.3
Presets Editor 1.0
Radio 1.0
Radio Now Playing 0.0.47
RadioNet 3.3
Random Mix 1.0
Ratings Light 2.8.4
Remote Music Libraries 1.0
Report Analytics Data 1.1
Rescan Music Library 1.0
RSS News Ticker 1.0
Save Playlist 1.0
Song Scanner 2.0
Sounds & Effects 1.1
Spotty 4.11.140
Switch Player - Group Player compatible 0.4
Virtual Library Creator 1.3.9
Visual Statistics 1.7.8
Visualizer Screensaver 1.0
What Was That Tune? 1.0.8
YouTube 0.300
```

### Is there relevant log output for this?

- [X] I have searched the server log

### Relevant log output

```shell
[24-12-18 18:16:16.8752] Slim::Control::Request::notify (2056) Error: Failed notify: Can't call method "id" on an undefined value at /var/lib/squeezeboxserver/cache/InstalledPlugins/Plugins/CustomSkip3/Plugin.pm line 1143.
[24-12-18 22:23:28.1771] Slim::Control::Request::notify (2056) Error: Failed notify: Can't call method "id" on an undefined value at /var/lib/squeezeboxserver/cache/InstalledPlugins/Plugins/CustomSkip3/Plugin.pm line 1130.
```

**Comment by mikeysas** [comment]
The following thread references a user who has also seen error msgs in their log after upgrading that are similar but referencing different line #'s

https://forums.slimdevices.com/forum/user-forums/logitech-media-server/1741693-lms-9-0-0-and-multiple-artists?p=1741767#post1741767

**Comment by AF-1** [comment]
I think it might be related to Material skin, among other things not saving the filter rule values correctly. CustomSkip like many other plugins uses html pages as UI. Material parses and changes them a bit.
Try this:
- enable the _Logic Teal_ or _Dark Logic Teal_ skin, then do **not** use _Material_ but the default url: yourlmsurloripaddress:9000 to enter the Custom Skip menu. 
- for each filter set go through every filter rule, check the values and change them if necessary, then **save** the filter rule (even if you haven't changed any values) 
- restart LMS. Then play a few tracks (maybe incl. some which would trigger a CustomSkip filter). 

I have checked this on my test setup with (Dark) Logic Teal and **no errors**. But I can see errors if I use Material skin.  Can you confirm that?

**Comment by mikeysas** [comment]
Yes, I think you are on to something regarding Material.  To test this, I first tried to make sure I could recreate the problem by creating a new filterset in Material which did not even seem to work correctly.  I was not getting errors in the log, but it was also not skipping.  So at this point I gave up on recreating the problem and just followed your instructions above.

I can confirm that I was able to force skips with Logic Teal after resaving the rule sets and not get any errors.  I am still having trouble recreating the errors in Material.  When I went into the rulesets in Logic Teal, some of them populated with the wrong values which I needed to correct and others were fine.  After saving them in Logic Teal, I have not been able to force the error again in Material???

Is that helpful or is there something else you would like me to try?

**Comment by AF-1** [comment]
I think something happended to selected values when they are saved with Material. Or maybe even before when Material creates the page.
After you have saved/corrected the values (correctly) with Logic Teal, CustomSkip can use them, so no more errors, at least not until you save them again with Material.
Maybe @CDrummond can take a look at what's happening.

**Comment by CDrummond** [comment]
Have you tested with "Classic" (**not** "Default")??? As all Material is adding some CSS to "Classic"

**Comment by mikeysas** [comment]
@CDrummond - I just ran a test with Classic and the Custom Skip plugin behaved as expected and no errors.  

I then replicated the same test using Material and Custom Skip filter rule did not skip the track.  It did not produce the same error message that I posted, but it is not functioning when the filter rule is saved in Material for some reason.  I have not been able to recreate the scenario that generates the error message yet.

Here are my replication steps in Material with the Custom Skip plugin installed
1) click Extras > Custom Skip
2) click Create New Filter Set
3) Enter a name and click Next
4) Select a genre to skip by expanding Genres and then click Genres
5) Choose the Genre you want to test to skip and click Save
6) Add tracks to queue that with multiple Genres including the one selected to skip starting at track 3.
7) hit play and it will progress through the tracks.

In Material it played through the playlist including the tracks for the selected Genre.  In Classic the selected Genre is skipped.

Additional observations:
- If you go back into the Genre Filter rule saved in Material, you will see that the selected Genre was not actually saved which I think it the problem.  In Classic, the selected Genre is saved.
- I have also seen intermittent issues where the edit & delete buttons are not visible on the Custom Skip filter set page like they are on Classic.  I think this happens after having save a new Filter Set using Material.  Also when those buttons were available, the contrast color of them is very hard to see on the Lyrion theme.

**Comment by CDrummond** [comment]
@mikeysas Thanks for that, I can confirm the issue is with Material. FYI, Material uses some 3rd party javascript to re-theme the combo-box style menus (where you select the Genre) and this appears to be breaking things here.

**Comment by CDrummond** [comment]
@AF-1 Please feel free to close this but, its a Material-specific issue and I've removed the offending code for the next Material release. Sorry for the hassle.

**Comment by AF-1** [comment]
Thank you all for looking into this and solving it.
@CDrummond one unrelated question: For existing filter sets there are 2 icons in the  CustomSkip menu: delete and edit. I must admit I haven't looked at this menu in Material in a long time. But these icons are barely visible now (see screenshot).
Did I inadvertently change a Material interface setting that does this? The edit icon is supposed to be an icon provided by LMS / Material while the svg delete icon comes with the plugin. They are both affected the same. 
Any idea how to make them more legible / visible ? Thanks.

![sc](https://github.com/user-attachments/assets/8c90d0c1-0c11-44b9-ade0-9fc547cd769f)

**Comment by CDrummond** [comment]
@AF-1 Ah, another Material issue!!! Fixed in git for 5.7.1, thanks for reporting.

**Comment by AF-1** [comment]
Thank you!
